File: dotnet/src/InstrumentComponents/Classes/Counter.cs
using InstrumentComponents.Dialects;
using InstrumentComponents.Identity;
using InstrumentComponents.Kind;
using InstrumentComponents.Session;
using InstrumentComponents.Scpi;

namespace InstrumentComponents.Classes;

/// <summary>Frequency counter session view (IVI-inspired / SCPI :MEASure, :COUNter).</summary>
public sealed class Counter : IInstrumentIdentity, IInstrumentShutdown
{
    private readonly InstrumentSession _session;

    public Counter(InstrumentSession session) => _session = session;

    public InstrumentSession Session => _session;

    public Idn QueryIdn() => _session.Idn();

    public void Reset() => _session.Reset();

    /// <summary>Counters have no output stage; safe-shutdown is a no-op.</summary>
    public void OutputOff()
    {
    }

    private DialectProfile Dialect => _session.DialectFor(InstrumentKind.Counter);

    private string Cmd(string key, string fallback, params (string Name, string Value)[] vars) =>
        DialectCommand.Try(Dialect, key, fallback, vars);

    public double MeasureFrequency() =>
        QueryF64(Cmd("measure_frequency", ScpiCommands.CounterMeasureFrequency));

    public double MeasurePeriod() =>
        QueryF64(Cmd("measure_period", ScpiCommands.CounterMeasurePeriod));

    public void SetGateTime(double seconds) =>
        _session.Scpi.Write(Cmd("gate_time", ScpiCommands.CounterGateTime(seconds),
            ("seconds", ScpiFormat.Double(seconds))));

    public void SelectChannel(uint channel) =>
        _session.Scpi.Write(Cmd("channel_select", ScpiCommands.CounterChannelSelect(channel),
            ("channel", channel.ToString())));

    public void ResetTotalize() =>
        _session.Scpi.Write(Cmd("reset_totalize", ScpiCommands.CounterResetTotalize));

    public double ReadTotalize() =>
        QueryF64(Cmd("read_totalize", ScpiCommands.CounterReadTotalize));

    private double QueryF64(string cmd) =>
        ScpiSession.ParseF64(_session.Scpi.Query(cmd));
}

File: crates/instrument/src/classes/counter.rs
use super::dialect_io;
use instrument_core::error::Result;
use instrument_core::kind::InstrumentKind;
use instrument_core::scpi::ScpiSession;
use instrument_core::scpi_commands;
use instrument_core::InstrumentSession;

/// Frequency counter session view (IVI-inspired / SCPI :MEASure, :COUNter).
pub struct Counter {
    session: InstrumentSession,
}

impl Counter {
    pub fn new(session: InstrumentSession) -> Self {
        Self { session }
    }

    pub fn session(&self) -> &InstrumentSession {
        &self.session
    }

    pub fn session_mut(&mut self) -> &mut InstrumentSession {
        &mut self.session
    }

    fn dialect(&self) -> &'static instrument_core::DialectProfile {
        self.session.dialect_for(InstrumentKind::Counter)
    }

    fn cmd(&self, key: &str, vars: &[(&str, String)], fallback: String) -> String {
        dialect_io::try_formatted(self.dialect(), key, vars, fallback)
    }

    /// Measures frequency in hertz (SI).
    pub fn measure_frequency(&mut self) -> Result<f64> {
        self.query_f64(dialect_io::try_command(
            self.dialect(),
            "measure_frequency",
            scpi_commands::COUNTER_MEASURE_FREQUENCY,
        ))
    }

    /// Measures period in seconds (SI).
    pub fn measure_period(&mut self) -> Result<f64> {
        self.query_f64(dialect_io::try_command(
            self.dialect(),
            "measure_period",
            scpi_commands::COUNTER_MEASURE_PERIOD,
        ))
    }

    /// Sets gate / aperture time in seconds (SI).
    pub fn set_gate_time(&mut self, seconds: f64) -> Result<()> {
        let cmd = self.cmd(
            "gate_time",
            &[("seconds", dialect_io::f64_text(seconds))],
            scpi_commands::counter_gate_time(seconds),
        );
        self.session.scpi_mut().write(&cmd)
    }

    /// Selects the measurement channel (1-based).
    pub fn select_channel(&mut self, channel: u32) -> Result<()> {
        let cmd = self.cmd(
            "channel_select",
            &[("channel", channel.to_string())],
            scpi_commands::counter_channel_select(channel),
        );
        self.session.scpi_mut().write(&cmd)
    }

    /// Resets the totalize counter.
    pub fn reset_totalize(&mut self) -> Result<()> {
        let cmd = dialect_io::try_command(
            self.dialect(),
            "reset_totalize",
            scpi_commands::COUNTER_RESET_TOTALIZE,
        );
        self.session.scpi_mut().write(cmd)
    }

    /// Reads totalize count.
    pub fn read_totalize(&mut self) -> Result<f64> {
        self.query_f64(dialect_io::try_command(
            self.dialect(),
            "read_totalize",
            scpi_commands::COUNTER_READ_TOTALIZE,
        ))
    }

    fn query_f64(&mut self, cmd: &str) -> Result<f64> {
        let resp = self.session.scpi_mut().query(cmd)?;
        ScpiSession::parse_f64(&resp)
    }
}
